File: level 2/기출문제/방금그곡/solution.py
def solution(m, musicinfos):
    m = m.replace("C#", "H").replace("D#", "I").replace("F#", "J").replace("G#", "K").replace("A#", "L")

    def subtract_time(a, b):
        a = a.split(':')
        b = b.split(':')
        a_min = int(a[0]) * 60 + int(a[1])
        b_min = int(b[0]) * 60 + int(b[1])
        return b_min - a_min

    answers = []

    for info in musicinfos:
        melody = ""
        new_info = info.split(',')
        new_info[3] = new_info[3].replace("C#", "H").replace("D#", "I").replace("F#", "J").replace("G#", "K").replace("A#", "L")
        time = subtract_time(new_info[0], new_info[1])
        for i in range(time):
            if i >= len(new_info[3]):
                i %= len(new_info[3])
            melody += new_info[3][i]
        if m in melody:
            if answers:
                if answers[0][1] < time:
                    answers[0] = [new_info[2], time]
            else:
                answers.append([new_info[2], time])
    if answers:
        return answers[0][0]
    else:
        return "(None)"

# 조금더 깔끔한 버전

def solution(m, musicinfos):
    m = m.replace('A#','H');
    m = m.replace('C#','I');
    m = m.replace('D#','J');
    m = m.replace('F#','K');
    m = m.replace('G#','L');
    answer = ''
    result = None
    dic = dict()
    for info in musicinfos:
        start, end, title, sound = info.split(',')
        hour1, min1 = start.split(':')
        hour2, min2 = end.split(':')
        time = (int(hour2)-int(hour1))*60 + int(min2)-int(min1)
        sound = sound.replace('A#','H');
        sound = sound.replace('C#','I');
        sound = sound.replace('D#','J');
        sound = sound.replace('F#','K');
        sound = sound.replace('G#','L');
        sound = sound*(time//len(sound))+sound[0:time%len(sound)]
        dic[sound] = title
    for song in dic.keys():
        if m in song:
            if result == None:
                result = song
            else:
                if len(result) < len(song):
                    result = song
    if result != None:
        return dic[result]
    else: return "(None)"

File: level 2/기출문제/방금그곡/solution.js
function solution(m, musicinfos) {
  const getMinutes = (s) => s.substr(0, 2) * 60 + +s.substr(3);

  m = m.replace(/\w#/g, (a) => a[0].toLowerCase());

  let answer = musicinfos
    .map((info) => {
      info = info.split(",");

      let playMinutes = getMinutes(info[1]) - getMinutes(info[0]),
        melody = info[3].replace(/\w#/g, (a) => a[0].toLowerCase());

      melody =
        playMinutes > melody.length
          ? melody.padEnd(playMinutes, melody)
          : melody.substr(0, playMinutes);

      return {
        playMinutes: playMinutes,
        melody: melody,
        name: info[2],
        startTime: getMinutes(info[0]),
      };
    })
    .filter((info) => new RegExp(m).test(info.melody));

  if (answer.length) {
    if (answer.length > 1) {
      answer = answer.filter(
        (v) =>
          v.playMinutes >= Math.max(...answer.map((val) => val.playMinutes))
      );

      if (answer.length > 1) {
        answer = answer.filter(
          (v) => v.startTime <= Math.min(...answer.map((val) => val.startTime))
        );
      }
    }

    return answer[0].name;
  }

  return "(None)";
}
